File: 002-Add-Two-Numbers/Solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    var carry = 0;
    var result = new ListNode(0);
    var ptr1 = l1;
    var ptr2 = l2;
    var val;
    var result_ptr = result;
    
    while(ptr1 && ptr2) {
        val = ptr1.val + ptr2.val + carry;
        carry = val > 9 ? 1 : 0;
        result_ptr.next = new ListNode(val % 10);
        result_ptr = result_ptr.next;
        ptr1 = ptr1.next;
        ptr2 = ptr2.next;
    }
    ptr1 = ptr1 ? ptr1 : ptr2;
    while(ptr1) {
        val = ptr1.val + carry;
        carry = val > 9 ? 1 : 0;
        result_ptr.next = new ListNode(val % 10);
        result_ptr = result_ptr.next;
        ptr1 = ptr1.next;
    }
    if(carry) {
        result_ptr.next = new ListNode(carry);
        result_ptr = result_ptr.next;
    }
    return result.next;
};

File: 002-Add-Two-Numbers/Solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *result, *ptr1, *ptr2, *result_ptr;
        int carry = 0;
        result = new ListNode(0);
        result_ptr = result;
        ptr1 = l1;
        ptr2 = l2;
        while(ptr1 && ptr2) {
            int val = ptr1->val + ptr2->val + carry;
            carry = val > 9 ? 1 : 0;
            result_ptr->next = new ListNode(val % 10);
            result_ptr = result_ptr->next;
            ptr1 = ptr1->next;
            ptr2 = ptr2->next;
        }
        ptr1 = ptr1 ? ptr1 : ptr2;
        while(ptr1) {
            int val = ptr1->val + carry;
            carry = val > 9 ? 1 : 0;
            result_ptr->next = new ListNode(val % 10);
            result_ptr = result_ptr->next;
            ptr1 = ptr1->next;
        }
        if(carry) {
            result_ptr->next = new ListNode(carry);
            result_ptr = result_ptr->next;
        }
        ptr1 = result;
        delete ptr1;
        result = result->next;
        return result;
    }
};
